### PR TITLE
Station's Technical Test

### PR DESCRIPTION
**Pull request by matthieukern**
# Start the project
To start the project, the commands remain the same. Clone the repo, then execute a `npm install` to install project dependencies. When it's done, run the project using `npm start -s` (or without the `-s` if you want to see the full output).

# What I implemented
I did not have time to implement the second version of the test. The only testable part is the first version: a simple chat with one room, that any user can join and exchange messages in, the user having access to the last 10 sent messages when connecting to the room.

When logged, to test the implemented features, you can use the `Chat` link in the header menu or navigate to the `/chat` route. This will display a chat, that should already be connected to the chatroom. The easiest way to test the chat is to open a second window, either using another browser or a private navigation tab, and start chatting with yourself!

The chat is very classic. When you click on send or hit enter it sends a message, when you use shift+enter it creates a new line. The email of the user that sent the message is printed on top of each message. Your messages are on the right part of the screen when your contacts messages are on the left side.

### CODE REVIEW

File: src/components/chat/Input.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    };
+  }
+
+  /** Helper to send a message using the state value. */
+  _sendMessage() {
+    const { message } = this.state;
+    if (message.trim().length > 0) {
+      this.props.sendMessage({
+        text: message,
+        sender: this.props.email
```
Comment (matthieukern) [inline src/components/chat/Input.js]: I did not implement any backend function actually, I had no time to look how Firebase is dealing with those kind of stuff.

File: src/components/chat/ChatRoom.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    }
+  }
+
+  render() {
+    return (
+      <div className="conversation">
+        {this.props.messages.map((message, key) => (
+          <div
+            key={key}
+            className={`message ${this.props.email === message.sender ? null : 'other'}`}
```
Comment (matthieukern) [inline src/components/chat/ChatRoom.js]: I never used Reselect, seems interesting!

Comment (alexstrat) [inline src/components/chat/ChatRoom.js]: Ya but even without reselect. The idea was to move the logic outside the component and only do presentation in components.

```js
getMessages = state => {
  const email = state.user.email;
  state.messages.map(msg => ({
    ...msg, senderIsMe: msg.email === email
  }));
}

File: src/components/chat/ChatRoom.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  scrollToEnd() {
+    if (this.anchor) {
+      this.anchor.scrollIntoView();
+    }
+  }
+
+  render() {
+    return (
+      <div className="conversation">
+        {this.props.messages.map((message, key) => (
```
Comment (alexstrat) [inline src/components/chat/ChatRoom.js]: Recommended to use a unique identifier rather an index when possible: https://facebook.github.io/react/docs/lists-and-keys.html#keys

File: src/components/chat/ChatRoom.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  componentWillMount() {
+    this.props.connectToChatRoom();
+  }
+
+  componentDidMount() {
+    this.scrollToEnd();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (prevProps.messages !== this.props.messages) {
```
Comment (alexstrat) [inline src/components/chat/ChatRoom.js]: Basically the same a before.
Imagine the case a message in the thread is edited or deleted: we will automatically scroll to the bottom line whereas that's not intended. I'd rather have checked if an element was added at the end.
But it's still a minor remark.

File: src/actions/messagesActions.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        dispatch(sendMessageFailure());
+      })
+    ;
+  };
+}
+
+export function connectToChatRoom() {
+  return dispatch => {
+    dispatch(connectedToChatRoom());
+    firebaseApi.SubscribeToValue('messages', message => dispatch(messageReceived(message.val())));
```
Comment (alexstrat) [inline src/actions/messagesActions.js]: I guess it'd simplify a lot of things if you'd extract the `key` of the message from firebase.

https://firebase.google.com/docs/database/web/lists-of-data

File: src/reducers/messagesReducer.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    case types.CHAT_MESSAGE_SENT_SUCCESS:
+      break;
+
+    case types.CHAT_MESSAGE_SENT_FAILURE:
+      break;
+
+    case types.CHAT_MESSAGE_RECEIVED:
+      newState = state.slice(0);
+      newState.push(action.message);
```
Comment (alexstrat) [inline src/reducers/messagesReducer.js]: Would have preferred a `key` based reducer.
_see comment on firebase's `key`_

File: src/components/chat/ChatRoom.js (hunk 3 above)
Comment (alexstrat) [inline src/components/chat/ChatRoom.js]: _see comment on firebase's `key`_

**Pull request closed without merging**